### PR TITLE
Feature: FROM-IP-CIDR rule type

### PR DESCRIPTION
**Pull request by songchenwen**
When running Clash in router, I want some of my clients to only go through a part of Clash's rules.

Specifically, I don't want my NAS to go through the GeoIP rule, since I run BT downloads in it. 

With this rule type and some changes in the rule sequence, Clash can be more flexible as a redirector in router.

### CODE REVIEW

**Comment by comzyh** [comment]
Eventually, clash becomes iptables, LOL. 
Actually, you can use iptables to filter source ip 
And, I suggest `source IP` rather than `from IP`

**Comment by songchenwen** [comment]
> Actually, you can use iptables to filter source ip
Yes, I used to do this with iptables. But it makes my configuration much more complicated.

> And, I suggest `source IP` rather than `from IP`
There is already a `source` in metadata, and it has a totally different meaning. It can be confusing to put `source ip` and `source` together.

File: rules/from_ipcidr.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+func NewFromIPCIDR(s string, adapter string) *FromIPCIDR {
+	_, ipnet, err := net.ParseCIDR(s)
+	if err != nil {
+	}
+	return &FromIPCIDR{
+		ipnet:   ipnet,
+		adapter: adapter,
+	}
+}
```
Comment (beyondkmp) [inline rules/from_ipcidr.go]: We can use ipcidr.go for fromipcidr rule.

Comment (songchenwen) [reply rules/from_ipcidr.go]: I figured out how to do this and pushed a new commit.

File: config/config.go [hunk 2]
```diff
@@ -338,7 +338,9 @@ func parseRules(cfg *rawConfig) ([]C.Rule, error) {
 		case "GEOIP":
 			rules = append(rules, R.NewGEOIP(payload, target))
 		case "IP-CIDR", "IP-CIDR6":
-			rules = append(rules, R.NewIPCIDR(payload, target))
+			rules = append(rules, R.NewIPCIDR(payload, target, false))
+		case "FROM-IP-CIDR":
```
Comment (Dreamacro) [inline config/config.go]: 叫 `SOURCE-CIDR` 或者 `SOURCE-IP-CIDR` 会不会好点

File: constant/metadata.go [hunk 3]
```diff
@@ -33,15 +33,20 @@ type SourceType int
...
+	FromIP   *net.IP
 	AddrType int
 	Host     string
 	IP       *net.IP
 	Port     string
 }
 
 func (addr *Metadata) String() string {
+	fromIP := ""
+	if addr.FromIP != nil {
```
Comment (Dreamacro) [inline constant/metadata.go]: 这里 FromIP 不可能为 nil

File: rules/ipcidr.go [hunk 4]
```diff
@@ -9,18 +9,29 @@ import (
 type IPCIDR struct {
 	ipnet   *net.IPNet
 	adapter string
+	isFrom  bool
 }
 
 func (i *IPCIDR) RuleType() C.RuleType {
-	return C.IPCIDR
+	if i.isFrom {
```
Comment (Dreamacro) [inline rules/ipcidr.go]: 更建议这么写

```go
if i.isFrom {
	return C.FromIPCIDR
} 
return C.IPCIDR
```

File: rules/ipcidr.go [hunk 5]
```diff
@@ -9,18 +9,29 @@ import (
...
+		return C.FromIPCIDR
+	} else {
+		return C.IPCIDR
+	}
 }
 
 func (i *IPCIDR) IsMatch(metadata *C.Metadata) bool {
-	if metadata.IP == nil {
+	var ip *net.IP
```
Comment (Dreamacro) [inline rules/ipcidr.go]: 这里和上面类似

File: rules/ipcidr.go [hunk 6]
```diff
@@ -9,18 +9,29 @@ import (
...
+	if i.isFrom {
+		ip = metadata.FromIP
+	} else {
+		ip = metadata.IP
+	}
+	if ip == nil {
```
Comment (Dreamacro) [inline rules/ipcidr.go]: 这里不需要判断

File: tunnel/tunnel.go [hunk 7]
```diff
@@ -106,6 +106,20 @@ func (t *Tunnel) handleConn(localConn C.ServerAdapter) {
...
+	case *InboundAdapter.HTTPAdapter:
+		if addr, ok := adapter.Conn().RemoteAddr().(*net.TCPAddr); ok {
+			ip = addr.IP
+		}
+	case *InboundAdapter.SocketAdapter:
+		if addr, ok := adapter.Conn().RemoteAddr().(*net.TCPAddr); ok {
+			ip = addr.IP
+		}
+	}
+	metadata.FromIP = &ip
```
Comment (Dreamacro) [inline tunnel/tunnel.go]: 处理 Source IP 不应该在这里，应该在 inbound proxy 里

**Comment by songchenwen** [comment]
@Dreamacro 改了 review 的内容并且 squash 了

File: adapters/inbound/http.go [hunk 8]
```diff
@@ -32,8 +32,10 @@ func (h *HTTPAdapter) Conn() net.Conn {
 
 // NewHTTP is HTTPAdapter generator
 func NewHTTP(request *http.Request, conn net.Conn) *HTTPAdapter {
+	metadata := parseHTTPAddr(request)
+	metadata.FromIP = parseFromIP(conn)
```
Comment (beyondkmp) [inline adapters/inbound/http.go]: 是不是代码里面的FromIP也改成SourceIP,统一下更好

File: constant/metadata.go [hunk 9]
```diff
@@ -33,15 +33,17 @@ type SourceType int
...
 	Source   SourceType
+	FromIP   *net.IP
 	AddrType int
 	Host     string
 	IP       *net.IP
 	Port     string
 }
 
 func (addr *Metadata) String() string {
+	fromIP := addr.FromIP.String() + " --> "
```
Comment (comzyh) [inline constant/metadata.go]: take care of this change.

https://github.com/Dreamacro/clash/blob/bd6c6a9ad1a79a677553cde7c8a956d1ce5b2c11/tunnel/tunnel.go#L137
https://github.com/Dreamacro/clash/blob/bd6c6a9ad1a79a677553cde7c8a956d1ce5b2c11/tunnel/tunnel.go#L160
https://github.com/Dreamacro/clash/blob/bd6c6a9ad1a79a677553cde7c8a956d1ce5b2c11/tunnel/tunnel.go#L164

Comment (songchenwen) [reply constant/metadata.go]: How to take care of this? I think when this rule type is added we need to show source ip in the log.

File: adapters/inbound/http.go (hunk 8 above)
Comment (songchenwen) [reply adapters/inbound/http.go]: 改好了

File: constant/metadata.go [hunk 10]
```diff
@@ -33,15 +33,17 @@ type SourceType int
...
 	NetWork  NetWork
 	Source   SourceType
+	SourceIP *net.IP
 	AddrType int
 	Host     string
 	IP       *net.IP
 	Port     string
 }
 
 func (addr *Metadata) String() string {
```
Comment (Dreamacro) [inline constant/metadata.go]: 这个函数不能这么写，因为 Dial TCP 那边需要用到这个函数，可以直接在 Tunnel 里拼

File: rules/ipcidr.go [hunk 11]
```diff
@@ -9,18 +9,22 @@ import (
 type IPCIDR struct {
 	ipnet   *net.IPNet
 	adapter string
+	isFrom  bool
```
Comment (Dreamacro) [inline rules/ipcidr.go]: 这个一起改了吧

File: constant/metadata.go (hunk 9 above)
Comment (comzyh) [reply constant/metadata.go]: As Dreamacro  said below

**Comment by songchenwen** [comment]
改了

**Pull request merged**